### ISSUE TITLE
WC 4.9.2+ (including 5.*) Breaks PSR Autoloading

### ISSUE THREAD

**Issue by rwrobe** [woocommerce/woocommerce#29011]
**Describe the bug**
I have a plugin that uses PSR-4 autoloading. Give a composer.json file with the following autoloader entry:
```
    "autoload": {
        "psr-4": {
            "MyPluginName\\": "public/content/plugins/mypluginname/src"
        }
    }
```
and `"woocommerce/woocommerce": "4.9.1",`, all plugin files load correctly. When I jump to `4.9.2`, autoloading breaks.

**Expected behavior**
Update to 4.9.2 with autoloading behaving as expected.

**Actual behavior**
Autoloader classMap does not list any classes that use Composer's autoload directive.

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
1. Update WC from 4.9.1 to 4.9.2
2. Composer up
3. Watch stuff break

**Isolating the problem (mark completed items with an [x]):**
- [x ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x ] I can reproduce this bug consistently using the steps above.

**Comment by juliaamosova** [comment]
Hi @rwrobe,

Thank you for submitting the issue. Can you please provide a bit more details of the issue? 

>Watch stuff break

Please specify what exactly breaks. Any error messages?

**Comment by rwrobe** [comment]
@juliaamosova Any classes you set up to resolve via autoloading are not found. This code works with 4.9.1:

```php


use MyPlugin\Services;
use Psr\Container\ContainerInterface;

// Load vendors.
if ( ! file_exists( ABSPATH . '../../vendor/autoload.php' ) ) {
	die;
}

require_once ABSPATH . '../../vendor/autoload.php';

add_action( 'plugins_loaded', static function (): void {
	Services::instance()->init( __FILE__ );
}, 1, 0 );
```

After 4.9.2: 
```php
Fatal error: Uncaught Error: Class 'Services' not found in /app/public/content/plugins/my-plugin/myplugin.php:47
```

**Comment by rwrobe** [comment]
Also saw WooCommerce Subscriptions (Version 3.0.12) break after the update to 4.9.2 and deactivating my plugin:

```
Fatal error: Uncaught Error: Call to undefined function wc_get_logger() in /app/public/content/plugins/woocommerce-subscriptions/includes/class-wcs-retry-manager.php:402
```

**Comment by imonteiro** [comment]
I'm getting the same issue, when change WC 4.9.1 to 4.92:

`Fatal error: Uncaught Error: Class 'WooCommerce' not found in ...\app\site\web\app\plugins\woocommerce\woocommerce.php on line 47`

**Comment by juliaamosova** [comment]
Hi @rwrobe and @imonteiro,

Thank you for your follow-up on this issue! Our apologies for the delayed response here. 

I believe the issue had been fixed by now. Would you please let us know if you still experience it? Thank you!

**Comment by rwrobe** [comment]
@juliaamosova still no dice. I've tried with WC 4.9.2 and 5.2.2 and get the following:

```
NOTICE: PHP message: PHP Fatal error:  Uncaught Error: Class 'WooCommerce' not found in /app/public/content/plugins/woocommerce/woocommerce.php:47
Stack trace:
#0 /app/public/content/plugins/woocommerce-gateway-paypal-express-checkout/includes/class-wc-gateway-ppec-plugin.php(278): WC()
#1 /app/public/content/plugins/woocommerce-gateway-paypal-express-checkout/includes/class-wc-gateway-ppec-plugin.php(176): WC_Gateway_PPEC_Plugin->_check_dependencies()
#2 /app/public/wp/wp-includes/class-wp-hook.php(292): WC_Gateway_PPEC_Plugin->bootstrap('')
#3 /app/public/wp/wp-includes/class-wp-hook.php(316): WP_Hook->apply_filters(NULL, Array)
#4 /app/public/wp/wp-includes/plugin.php(484): WP_Hook->do_action(Array)
#5 /app/public/wp/wp-settings.php(423): do_action('plugins_loaded')
#6 /app/public/wp-config.php(9): require_once('/app/public/wp/...')
#7 /app/public/wp/wp-load.php(42): require_once('/app/public/wp-...')
#8 /app/public/wp/wp-admin/admin-ajax.php(22): require_once('/app/public/wp/...')
#9 {main}
  thrown in /app/public/content/plugins/woocommerce/woocommerce.php on line 47
```

The composer changes:

```
Downgrading automattic/jetpack-autoloader (v2.9.1 => v2.7.1): Loading from cache
  - Downgrading woocommerce/woocommerce-blocks (v4.7.2 => v4.0.0): Loading from cache
  - Downgrading woocommerce/woocommerce-admin (2.1.5 => 1.8.3): Loading from cache
  - Downgrading woocommerce/woocommerce (5.2.2 => 4.9.2): Downloading (100%)         
  - Updating composer/composer (1.10.21 => 1.10.22): Downloading (100%) 
 ```

**Comment by imonteiro** [comment]
@juliaamosova I've tried but without success.

**Comment by juliaamosova** [comment]
Hi @rwrobe and @imonteiro,

Thank you for checking again. 

This issue requires further feedback from the WooCommerce Core team. I am adding the `needs developer feedback` label to this issue so that the Core team could take a look. 

Please note it may take a few days for them to get to this issue. Thank you for your patience.

**Comment by rwrobe** [comment]
Hi, any progress on this issue? Thank you!

**Comment by szepeviktor** [comment]
@rwrobe Please see packages usage. (That is basically only me)
https://packagist.org/packages/woocommerce/woocommerce/stats
So now you can answer your own question!

**Comment by rwrobe** [comment]
@szepeviktor sorry, I'm not sure what information I should be getting from that or what you mean by your comment.

**Comment by lkraav** [comment]
> Sorry, I'm not sure what information I should be getting from that or what you mean by your comment.

He was trying to say 4.9 releases won't be getting any fixes.

You should try installing a newer version (any of them), and see if it has been improved by now, intentionally or unintentionally.

**Comment by rwrobe** [comment]
Thanks @lkraav -- I have this same issue in all versions of 5.* I have tested, so the title of the ticket probably should have been "since 4.9.2".

**Comment by lkraav** [comment]
Is there some service for PHP apps like https://glitch.me and https://webcomponents.dev for JS - where you could set up a demonstration of this issue for people to explore?

Usually the biggest barrier to investigation is trying to figure out how to replicate it.

**Comment by rwrobe** [comment]
So the response is that no fix will come because the official `woocommerce/woocommerce` Composer repository is unsupported?

>Basically no one uses WC through Composer, so do not expect a fix.

I have worked for quite a few agencies, including some on the WordPress VIP list. I assure you that every one of them uses composer to install and manage WooCommerce.

**Comment by rwrobe** [comment]
The fix I got working is to use the WPackagist repo. Change

`woocommerce/woocommerce` to `wpackagist-plugin/woocommerce` and remove `automattic/jetpack-autoloader` as a peer dependency.

**Comment by jeffstieler** [comment]
Closing since you found a solution. We will revisit the documentation for our Composer package to hopefully avoid confusion in the future.
